diff --git a/docroot/sites/all/modules/contrib/lang_dropdown/lang_dropdown.module b/docroot/sites/all/modules/contrib/lang_dropdown/lang_dropdown.module
index 34cdb61..afdaa46 100644
--- a/docroot/sites/all/modules/contrib/lang_dropdown/lang_dropdown.module
+++ b/docroot/sites/all/modules/contrib/lang_dropdown/lang_dropdown.module
@@ -622,6 +622,40 @@ function lang_dropdown_form($form, &$form_state, $languages, $type) {
   ($settings['languageicons']['flag_position']) ? $flag_position = '#suffix' : $flag_position = '#prefix';
 
   // Now we build the $form array.
+
+  // order languages alphabetically - by native language name
+  setlocale(LC_COLLATE, array(
+    'bg_BG.UTF-8',
+    'cs_CZ.UTF-8',
+    'da_DK.UTF-8',
+    'de_DE.UTF-8',
+    'el_CY.UTF-8',
+    'el_GR.UTF-8',
+    'en.UTF-8',
+    'en_US.UTF-8',
+    'es_ES.UTF-8',
+    'et_EE.UTF-8',
+    'fi_FI.UTF-8',
+    'fr_FR.UTF-8',
+    'hr_HR.UTF-8',
+    'hu.UTF-8',
+    'is_IS.UTF-8',
+    'it_IT.UTF-8',
+    'lt_LT.UTF-8',
+    'lat.UTF-8',
+    'mt_MT.UTF-8',
+    'nl_NL.UTF-8',
+    'no_NO.UTF-8',
+    'pl.UTF-8',
+    'pt_PT.UTF-8',
+    'ro_RO.UTF-8',
+    'sk_SK.UTF-8',
+    'sl_SI.UTF-8',
+    'sv_SE.UTF-8',
+    'sr_CS.UTF-8'
+  ));
+  asort($options,SORT_LOCALE_STRING);
+
   $form['lang_dropdown_select'] = array(
     '#type' => 'select',
     '#default_value' => isset($selected_option) ? $selected_option : key($options),
